File: android/java/org/chromium/chrome/browser/BraveFeatureList.java
/* Copyright (c) 2019 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;

@JNINamespace("chrome::android")
public abstract class BraveFeatureList {
    public static final String BRAVE_REWARDS = "BraveRewards";
    public static final String NATIVE_BRAVE_WALLET = "NativeBraveWallet";
    public static final String USE_DEV_UPDATER_URL = "UseDevUpdaterUrl";
    public static final String FORCE_WEB_CONTENTS_DARK_MODE = "WebContentsForceDark";
    public static final String ENABLE_FORCE_DARK = "enable-force-dark";
    public static final String ENABLE_TAB_GROUPS = "enable-tab-groups";
    public static final String ENABLE_TAB_GRID = "enable-tab-grid-layout";

    public static void enableFeature(
            String featureName, boolean enabled, boolean fallbackToDefault) {
        BraveFeatureListJni.get().enableFeature(featureName, enabled, fallbackToDefault);
    }

    @NativeMethods
    interface Natives {
        void enableFeature(String featureName, boolean enabled, boolean fallbackToDefault);
    }
}

File: browser/android/brave_feature_list.cc
/* Copyright (c) 2020 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 3.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include <string>

#include "base/android/jni_string.h"
#include "base/feature_list.h"
#include "brave/browser/brave_browser_process_impl.h"
#include "brave/build/android/jni_headers/BraveFeatureList_jni.h"
#include "chrome/browser/about_flags.h"
#include "components/flags_ui/pref_service_flags_storage.h"

namespace chrome {
namespace android {

static void JNI_BraveFeatureList_EnableFeature(
    JNIEnv* env,
    const base::android::JavaParamRef<jstring>& featureName,
    jboolean enabled,
    const base::android::JavaParamRef<jstring>& disabledValue) {
  std::string feature_name = ConvertJavaStringToUTF8(env, featureName);
  std::string disabled_value = ConvertJavaStringToUTF8(env, disabledValue);
  enabled ? feature_name += "@1" : feature_name += "@" + disabled_value;
  flags_ui::PrefServiceFlagsStorage flags_storage(
      g_brave_browser_process->local_state());
  about_flags::SetFeatureEntryEnabled(&flags_storage, feature_name, true);
}

}  // namespace android
}  // namespace chrome
